### ISSUE TITLE
Basic "scheduler" for core

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#295]
This is not meant to subsume sched, but rather to serve as the base implementation of the job submission interface being defined in #268, providing "run now or fail/block" functionality in flux-core.

"Scheduler" components for core:

* resource graph library and/or service:
  * ingestion/parsing
  * interface for queries
* Job submission interface
* Job feasibility check module, may check with the scheduler but need not be part of it
* Basic, probably blocking, FCFS "scheduler" with oversubscription avoidance and probably nothing else

**Comment by grondo** [comment]
I think we should bubble this one up. 

Would it be acceptable to have a basic, bare minimum sched in core that does fcfs scheduling using resource-hwloc, and can be used in testing within flux-core for anything that expects the base scheduler interfaces to be present (e.g. job.create, flux-submit, job specification lint, etc.)?

I would think this would be quick to implement and would not require advanced functionality that is going into the full flux-sched scheduler, and once it is complete there would be very little code churn.

**Comment by trws** [comment]
That sounds like exactly what this was meant to be about.  It would 
still have to be able to accept the same kind of job spec, but it 
wouldn't have to be able to do any kind of fancy scheduling, no 
backfill, no nothing, just find free resources and pack them in.

**Comment by dongahn** [comment]
This sounds great to me.

**Comment by lipari** [comment]
I guess I'd rather see flux-sched's rdl, resrc, and sched components migrated into flux-core.  The sched project was designed to support multiple plugins that include FCFS scheduling (currently implemented in schedplugin1).  This would allow us to work off a single flux-submit.  And it would eliminate [Issue 56](https://github.com/flux-framework/flux-sched/issues/56).  But mostly it would save us from implementing the same functionality in two places.

**Comment by trws** [comment]
That would be ideal as far as I'm concerned.  It's probably worthwhile to keep keep the un-core functionality, simulator and some test support code, separate but having sched in core where we can integrate it into check test it and depend on a version of it existing would be a **huge** win in my book.

**Comment by dongahn** [comment]
If we were to take this route, we will probably want to clearly define what part of sched belongs to core and what part to un-core as soon as possible and do the surgery. 

I guess the scheduler framework service would be a part of the core along with some basic scheduler plugin(s), rdl,  resrc and some utilities like waitjob. This will allow other external projects to write their own custom scheduling plugins to the scheduler framework service. In fact, @SteVwonder's IO-aware scheduler plugins will be the first such external project.  

These should be relatively easily separable at the moment. But when @SteVwonder completes his emulation environment merged into the framework service, there will be emulation-mode support within the framework service. Having the basic emulation support within "core" and keeping the rest of emulation mechanics uncore should be beneficial IMHO. But nevertheless, we should talk things through.

**Comment by garlick** [comment]
I second @dongahn's hesitation about merging all or a large chunk of flux-sched into flux-core.  The current split does allow flux-sched to be developed more or less independently of flux-core, and presumably in a more agile manner than we've settled in on with flux-core.  Clearly we need better integration but perhaps the approach outlined in this issue is sufficient for the time being?

**Comment by dongahn** [comment]
@garlick: I have a mixed feeling about this, but I do think it is a great idea to carve out a part of sched into the core (in particular the framework service part). There are lot of benefits with tigher integration with core, IMHO. So it seems this begs a question, when is the good time if it is not now and how we can grow sched so that in the end it can be separable. If this is not the right time, I feel we need to define a future milestone that will steer the development of sched into a right direction.

**Comment by grondo** [comment]
Yeah, I think the idea would be to have a basic sched module in flux-core which
 - hardly ever needs to change
 - when it does need to change the code is small and simple enough for even 3rd party developer could make necessary mods and it will be easy to write tests with good coverage

Keeping the sched module in core to a minimum seems like the most straightforward approach for now.

**Comment by trws** [comment]
Fair enough, so let's define what things are currently considered part of the scheduler that need to be in core, and what they need to do to get us where we need to be.  I'm going to update the top-level issue here and put in a proposed checklist with a breakdown, if we agree on the items I'll break it into sub-issues.

**Comment by grondo** [comment]
Thanks @trws. I propose the tasks as follows:

 - augment current job module in core (actually replace it, nothing there worth saving) to have job.submit or similar interface as you suggested.
 - develop a scheme for interested entities to register with job module to act as filters on submit. Filters should be able to reject, amend, and/or add warnings to submit response.
 - develop a small scheduler that *only* uses resource-hwloc info and schedules fcfs, non-overlapping resources based on job.submit requests. I estimate this mini-sched to be just a few hundred lines (?) and will serve as simple test of interfaces above
 - feasibility check can be done by this sched or a very small bit of code that checks top-level info in resource-hwloc

This should give the most minimal set of services to core (simple as possible, but no simpler principle) for testing and development of these interfaces, and allow advanced scheduling work to continue on flux-sched.

**Comment by trws** [comment]
Agreed.  Could you tackle the job module, just assume a string blob for spec properties, and I'll look into doing the basic scheduler? I need to prototype something to handle the job spec syntax for core anyway, and that will be a non-trivial portion of it, it may also be able to lean on the hwloc code to do its layouts...

Anyway, the only point of contention I see might be the interface to actually submitting a job to the scheduler. Maybe just offer a way to order the filters you mentioned, and make a valid response "consumed?"  That way the scheduler just uses the same interface as everything else, but informs the system that it has accepted and taken responsibility for it, so don't pass it to any more filters?

**Comment by grondo** [comment]
Meant to reply that yes, I'll work on something for the job module. Design is a bit fluid at this point so I'll try to start a discussion over in #338

**Comment by trws** [comment]
This is no longer really an issue, since flux-sched is packaged and can be auto-loaded.
